GVFS/GVFS.Service/ProductUpgradeTimer.cs

```diff
@@ -55,28 +56,37 @@ public void Dispose()
         private void TimerCallback(object unusedState)
         {
             string errorMessage = null;
-
             InstallerPreRunChecker prerunChecker = new InstallerPreRunChecker(this.tracer, string.Empty);
-            ProductUpgrader productUpgrader = new ProductUpgrader(ProcessHelper.GetCurrentProcessVersion(), this.tracer);
-            if (prerunChecker.TryRunPreUpgradeChecks(out string _) && this.TryDownloadUpgrade(productUpgrader, out errorMessage))
+            IProductUpgrader productUpgrader;
+
+            if (ProductUpgrader.TryCreateUpgrader(out productUpgrader, this.tracer, out errorMessage))
             {
-                return;
-            }
+                if (prerunChecker.TryRunPreUpgradeChecks(out string _) && this.TryDownloadUpgrade(productUpgrader, out errorMessage))

```

Comment:
Can\should `TryRunPreUpgradeChecks` be checked before creating an upgrader (`TryCreateUpgrader`)?

Reply:
`TryCreateUpgrader` checks if the upgrade configuration is set correctly. It helps early exit with error (when config is incorrect) or info (when config is set to disable upgrade) message. The sequence of error/info messages that get displayed has gone through multiple rounds of reviews. The sequence is the same in `UpgradeVerb`, `UpgraderTool` and in `GVFS.Service` Upgrade timer. If we change the order to do `TryRunPreUpgradeChecks` first, then this message sequence would change as well. 

Example: After a fresh install of VFSForGit, `gvfs upgrade` (or service timer) might display pre-check specific error and exit, which might be incorrect because the User has not yet specified upgrade config (in this case the better behavior would be to give instruction to user on how to set upgrade config.)